src/gui/addnewtorrentdialog.cpp

```diff
@@ -779,19 +775,39 @@ void AddNewTorrentDialog::handleDownloadFailed(const QString &url, const QString
 void AddNewTorrentDialog::handleRedirectedToMagnet(const QString &url, const QString &magnetUri)
 {
     Q_UNUSED(url)
-    if (loadMagnet(BitTorrent::MagnetUri(magnetUri)))
-        open();
-    else
+
+    if (!loadMagnet(BitTorrent::MagnetUri(magnetUri))) {
         this->deleteLater();
+        return;
+    }
+
+#ifdef Q_OS_MAC
+    setModal(true);
+#endif
+    open();
 }
 
-void AddNewTorrentDialog::handleDownloadFinished(const QString &url, const QString &filePath)
+void AddNewTorrentDialog::handleDownloadFinished(const QString &url, const QByteArray &data)
 {
-    Q_UNUSED(url)
-    if (loadTorrent(filePath))
-        open();
-    else
+    QString error;
+    const BitTorrent::TorrentInfo info = BitTorrent::TorrentInfo::load(data, &error);
+    if (!info.isValid()) {
+        RaisedMessageBox::critical(this, tr("Invalid torrent"), tr("Failed to load from URL: %1.\nError: %2")
+            .arg(url, error));
+        return;
+    }
+
+    m_torrentGuard.reset(new TorrentFileGuard);
+
+    if (!loadTorrentImpl(info)) {
         this->deleteLater();
+        return;
+    }
+
+#ifdef Q_OS_MAC
```

Comment:
```c++
#ifdef Q_OS_MAC
    setModal(true);
#endif
    open();
```
This code part is repeated 3 times. It seems that it should be extracted into a separate method.

Reply:
Sure, done.